src/main/java/org/opennars/inference/CompositionalRules.java

```diff
@@ -756,9 +757,23 @@ static void IntroVarSameSubjectOrPredicate(final Sentence originalMainSentence,
             }
         }
         
-        Set<Set<Term>> powerset = powerSet(app.keySet());
+        List<Term> shuffledVariables = new ArrayList<Term>();
+        for(Term t : app.keySet()) {
+            shuffledVariables.add(t);
+        }
+        Collections.shuffle(shuffledVariables);
+        HashSet<Term> selected = new HashSet<Term>();
+        int i = 1;
+        for(Term t : shuffledVariables) {
+            selected.add(t);
+            if(Math.pow(2.0, i) > nal.narParameters.VARIABLE_INTRODUCTION_COMBINATIONS_MAX) {
```

Comment:
final double log10MaxCombinations = Math.log(nal.narParameters.VARIABLE_INTRODUCTION_COMBINATIONS_MAX) / Math.log(2);

if((double)i > log10MaxCombinations) 

is indeed better because there is no danger of overflow